File: go/0337M-打家劫舍III.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-02 16:47:05
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func rob(root *TreeNode) int {
	// 长度 2 的数组：0 不偷 1 偷
	var robTree func(node *TreeNode) []int
	robTree = func(node *TreeNode) []int {
		if node == nil {
			return []int{0, 0}
		}
		// 后序
		left := robTree(node.Left)
		right := robTree(node.Right)
		// 不偷
		val1 := max(left[0], left[1]) + max(right[0], right[1])
		// 偷
		val2 := node.Val + left[0] + right[0]
		return []int{val1, val2}
	}
	ans := robTree(root)
	return max(ans[0], ans[1])
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

File: cpp/0337M-打家劫舍III.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-12 19:37:33
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int rob(TreeNode* root) {
        vector<int> ans = robTree(root);
        return max(ans[0], ans[1]);
    }

    vector<int> robTree(TreeNode* root) {
        if (root == nullptr) return {0, 0};
        vector<int> left = robTree(root->left);
        vector<int> right = robTree(root->right);
        int v1 = max(left[0], left[1]) + max(right[0], right[1]);
        int v2 = left[0] + right[0] + root->val;
        return {v1, v2};
    }
};